### PR TITLE
GNOI Cold Reboot

### PR DESCRIPTION
**Pull request by rkavitha-hcl**
##### Work item tracking
- Microsoft ADO **(number only)**:

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: src/sonic-framework/debian/control [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+Maintainer: Runming Wu <runmingwu@google.com>
+Section: net
+Priority: optional
+Build-Depends: dh-exec (>=0.3), debhelper (>= 9), autotools-dev
+Standards-Version: 1.0.0
+
+Package: framework
+Architecture: any
+Depends: ${shlibs:Depends}
+Description: This package contains framework service for GPINs project.
```
Comment (kishanps) [inline src/sonic-framework/debian/control:11]: Remove for GPINs project.

File: src/sonic-framework/gnoi/.bazelversion [hunk 2]
```diff
@@ -0,0 +1 @@
+6.4.0
```
Comment (kishanps) [inline src/sonic-framework/gnoi/.bazelversion:1]: The entire gnoi dir should show here as a sub-module

File: src/sonic-framework/rebootbackend/init_thread.cpp [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+#include "subscriberstatetable.h"
+#include "warm_restart.h"
+
+namespace rebootbackend {
+
+using WarmStartState = ::swss::WarmStart::WarmStartState;
+//using WarmBootStage = ::swss::WarmStart::WarmBootStage;
+
+InitThread::InitThread(
+//		       CriticalStateInterface &critical_interface,
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/init_thread.cpp:35]: Remove all commented code related to CriticalSate etc.

File: src/sonic-framework/rebootbackend/init_thread.cpp [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+    do_final_failed_actions();
+    m_finished.notify();
+    return;
+  }
+
+  m_status.set_detailed_thread_status(
+      InitThreadStatus::ThreadStatus::WAITING_FOR_RECONCILIATION);
+  // Reconciliation start time is recorded by the platform layer when the
+  // database is started.
+
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/init_thread.cpp:120]: Remove all commented code

File: src/sonic-framework/rebootbackend/init_thread.cpp [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+        "State verification timed out. Treating as success for NSF");
+    return swss::StatusCode::SWSS_RC_SUCCESS;
+  } else if (result != swss::StatusCode::SWSS_RC_SUCCESS) {
+    SWSS_LOG_ERROR("Error while waiting for state verification: %s",
+                   swss::statusCodeToStr(result).c_str());
+  }
+  return result;
+} */
+
+/* swss::StatusCode InitThread::select_loop(
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/init_thread.cpp:484]: Why is this commented out ?

File: src/sonic-framework/rebootbackend/init_thread.h [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+#include "subscriberstatetable.h"
+#include "system/system.pb.h"
+#include "warm_restart.h"
+
+namespace rebootbackend {
+
+// Holds a thread safe representation of the InitThread internal state.
+// Thread-safe: the expectation is one thread will write and multiple threads
+//   will read.
+class InitThreadStatus {
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/init_thread.h:25]: Comment for John: Is this needed, given we are not doing NSF orchestration ?

File: src/sonic-framework/rebootbackend/init_thread.h [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+  // Request InitThread stop/exit. Notifies the internal thread that it must
+  // exit. Only used when platform is shutting down all containers/processes.
+  virtual void Stop(void);
+
+  // Must be called by launching task after notification is sent to m_finished.
+  virtual bool Join(void);
+
+  // Return Status of last reboot attempt.
+  virtual gnoi::system::RebootStatusResponse GetResponse();
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/init_thread.h:159]: Have we tested RebootStatusResponse ? John, any inputs on how to test this ?

File: src/sonic-framework/rebootbackend/init_thread.h [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+  // otherwise.
+  swss::StatusCode handle_registration_step(
+      swss::SelectableTimer &timer_select);
+
+  // Helper function for the reconciliation step. Wait for all apps to reach
+  // the reconcilied state. The provided timer must already have been started
+  // prior to this function call.
+  // Returns SWSS_RC_SUCCESS if waiting for reconciliation is successful, and an
+  // error otherwise.
+  swss::StatusCode handle_reconciliation_step(
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/init_thread.h:199]: registration, reconcilation, freeze, unfreeze are not required.

File: src/sonic-framework/rebootbackend/init_thread.h [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+  // call.
+  swss::StatusCode wait_for_state(
+  //				swss::WarmStart::WarmBootStage nsf_stage,
+                                  swss::SelectableTimer &timer_select);
+
+  // Helper function for the state verification step. Trigger state verification
+  // then wait for all apps to report successful state verification.
+  // Returns SWSS_RC_SUCCESS if waiting for state verification is successful,
+  // and an error otherwise.
+  swss::StatusCode handle_state_verification_step();
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/init_thread.h:219]: Not required

File: src/sonic-framework/rebootbackend/interfaces.h [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+      : DBus::ObjectProxy(connection, dbus_obj_name_p, dbus_bus_name_p) {}
+};
+
+class HostServiceDbus : public DbusInterface {
+ public:
+  DbusInterface::DbusResponse Reboot(
+      const std::string& json_reboot_request) override;
+  DbusInterface::DbusResponse RebootStatus(
+      const std::string& json_status_request) override;
+  DbusInterface::DbusResponse StopContainers(
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/interfaces.h:37]: John: Is StopContainer needed ?

File: src/sonic-framework/rebootbackend/reboot_interfaces.h [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+  virtual ~DbusInterface() = default;
+  virtual DbusResponse Reboot(const std::string& json_reboot_request) = 0;
+  virtual DbusResponse RebootStatus(const std::string& json_status_request) = 0;
+  virtual DbusResponse StopContainers(const std::string& json_stop_request) = 0;
+  virtual DbusResponse StopContainerStatus(
+      const std::string& json_status_request) = 0;
+};
+
+class CriticalStateInterface {
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_interfaces.h:27]: Remove

File: src/sonic-framework/rebootbackend/reboot_thread.cpp [hunk 12]
```diff
@@ -0,0 +1,965 @@
+#include "reboot_thread.h"
+
+#include <google/protobuf/util/json_util.h>
+
+#include <chrono>
+
+//#include "component_state_helper.h"
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.cpp:7]: Remove

File: src/sonic-framework/rebootbackend/reboot_thread.cpp [hunk 13]
```diff
@@ -0,0 +1,965 @@
...
+  SWSS_LOG_NOTICE(
+      "done all pre-reboot steps, sending reboot request to platform");
+  if (send_dbus_reboot_request() == Progress::EXIT_EARLY) {
+    return Progress::EXIT_EARLY;
+  }
+
+  // Wait for platform to reboot. If we return, reboot failed.
+  // Logging, error status and monitoring for critical state are handled within.
+  return wait_for_platform_reboot(s);
+} */
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.cpp:207]: Remove

File: src/sonic-framework/rebootbackend/reboot_thread.cpp [hunk 14]
```diff
@@ -0,0 +1,965 @@
...
+
+void RebootThread::do_nsf_reboot(swss::Select &s) {
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.cpp:209]: This is not required

File: src/sonic-framework/rebootbackend/reboot_thread.cpp [hunk 15]
```diff
@@ -0,0 +1,965 @@
...
+    // is retriable.
+    log_error_and_set_non_retry_failure(
+        "state verification failed during reboot");
+    return Status::FAILURE;
+  } */
+
+  return Status::KEEP_WAITING;
+}
+
+Progress RebootThread::state_verification_select(
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.cpp:434]: remove

File: src/sonic-framework/rebootbackend/reboot_thread.cpp [hunk 16]
```diff
@@ -0,0 +1,965 @@
...
+        // handle_state_verification_event.
+        return Progress::EXIT_EARLY;
+      } else {
+        continue;
+      }
+    }
+  }
+}
+
+/* Progress RebootThread::perform_state_verification(swss::Select &s) {
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.cpp:478]: remove

File: src/sonic-framework/rebootbackend/reboot_thread.cpp [hunk 17]
```diff
@@ -0,0 +1,965 @@
...
+//
+// Stop On Freeze Support
+//
+
+// Perform quiescence and container stop in parallel.
+// First we request container stop.
+// Freeze is sent to all containers.
+// We wait for containers to quiesce (or checkpoint).
+// We wait for containers to stop.
+Progress RebootThread::perform_freeze_w_container_stop(swss::Select &s) {
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.cpp:520]: remove

File: src/sonic-framework/rebootbackend/reboot_thread.h [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+  // Args:
+  //   sub: [input] selectable subscription to STATE_VERIFICATION_RESP_TABLE
+  //        data is pending
+  //   timestamp: [input && output] the timestamp sent to state verification
+  //              with the state verification request
+  // Returns:
+  //   KEEP_WAITING: keep waiting for success, fail or timeout
+  //   SUCCESS: state verification passed, procced
+  //   FAILURE: state verification failed
+  Status handle_state_verification_event(swss::SubscriberStateTable &sub,
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.h:215]: Remove

File: src/sonic-framework/rebootbackend/reboot_thread.h [hunk 19]
```diff
@@ -0,0 +1,418 @@
...
+                                     swss::SelectableTimer &l_timer,
+                                     swss::SubscriberStateTable &sub,
+                                     std::string &timestamp);
+
+  // If enabled: perform non-frozen state verification
+  //   Check for critical state, listen for stop, handle a timeout
+  // Returns:
+  //   EXIT_EARLY: an issue occurred that stops NSF
+  //   PROCEED: if reboot timeout expired
+  Progress perform_state_verification(swss::Select &s);
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/reboot_thread.h:239]: remove

File: src/sonic-framework/rebootbackend/rebootbackend.cpp [hunk 20]
```diff
@@ -0,0 +1,16 @@
+#include "interfaces.h"
+#include "reboot_interfaces.h"
+#include "rebootbe.h"
+#include "telemetry_helper.h"
+
+using namespace ::rebootbackend;
+
+int main(int argc, char** argv) {
+  HostServiceDbus dbus_interface;
+  //CriticalState critical_interface;
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/rebootbackend.cpp:10]: Rmove commented lines

File: src/sonic-framework/rebootbackend/redis_utils.cpp [hunk 21]
```diff
@@ -0,0 +1,476 @@
...
+    m_missing_registrations.erase(key);
+  }
+}
+
+std::string InitRegistration::join_pending_apps() {
+  std::ostringstream stream;
+  std::copy(m_missing_registrations.begin(), m_missing_registrations.end(),
+            std::ostream_iterator<std::string>(stream, ","));
+  return stream.str();
+}
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/redis_utils.cpp:474]: Remove all wam reboot code

File: src/sonic-framework/rebootbackend/redis_utils.h [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+  // Pops the app name in key from the internal set, if it exists.
+  void remove_pending_app(const std::string &key);
+
+  RegistrationSet m_missing_registrations;
+  swss::DBConnector m_db;
+  std::string m_separator;
+
+  friend class RedisTest;
+};
+
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/redis_utils.h:289]: Remove all warm reboot code

File: src/sonic-framework/rebootbackend/telemetry_helper.cpp [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+/*void TelemetryHelper::clear_performance_table() {
+  swss::Table perf_table(&m_db, STATE_WARM_RESTART_PERF_TABLE_NAME);
+  std::vector<std::string> perf_keys;
+  perf_table.getKeys(perf_keys);
+  for (const auto &key : perf_keys) {
+    perf_table.del(key);
+  }
+} */
+
+}  // namespace rebootbackend
```
Comment (kishanps) [inline src/sonic-framework/rebootbackend/telemetry_helper.cpp:120]: John: Is Telemetry interface required ?

File: src/sonic-framework/tests/Makefile.am [hunk 24]
```diff
@@ -0,0 +1,60 @@
+INCLUDES = -I $(top_srcdir) -I $(top_srcdir)/rebootbackend -I $(top_srcdir)/rebootbackend/system
```
Comment (kishanps) [inline src/sonic-framework/tests/Makefile.am:1]: Have we run the tests ?

File: src/sonic-framework/rebootbackend/telemetry_helper.cpp (hunk 23 above)
Comment (johnham-google) [reply src/sonic-framework/rebootbackend/telemetry_helper.cpp:120]: Could you clarify/be more specific?

Comment (github76543) [reply src/sonic-framework/rebootbackend/telemetry_helper.cpp:120]: Can you be more specific? I don't know what you mean by Telemetry interface.

File: src/sonic-framework/rebootbackend/init_thread.h (hunk 6 above)
Comment (github76543) [reply src/sonic-framework/rebootbackend/init_thread.h:25]: InitThreadStatus is needed. InitThread itself isn't.

File: src/sonic-framework/rebootbackend/init_thread.h (hunk 7 above)
Comment (github76543) [reply src/sonic-framework/rebootbackend/init_thread.h:159]: I don't have good suggestions here: I'm unfamiliar with availble options.

File: src/sonic-framework/rebootbackend/interfaces.h (hunk 10 above)
Comment (github76543) [reply src/sonic-framework/rebootbackend/interfaces.h:37]: No: but I suggest leaving it in interfaces. Just don't call it when NSF warm boot code is stripped out.

**Comment by rkavitha-hcl** [comment]
'/azpw ms_conflict'

**Pull request closed without merging**